gui/unit-syndromes.lua

```diff
@@ -0,0 +1,543 @@
+        if unit.animal.population.feature_idx == -1 and unit.animal.population.cave_id == -1 then
+            if dfhack.units.isAnimal(unit) and not dfhack.units.isFortControlled(unit) then
+                table.insert(units, unit)
+            end
+        end
+    end
+
+    return units
+end
+
+local function getHostiles()
+    local units = {}
+
+    for _, unit in pairs(df.global.world.units.active) do
+        if dfhack.units.isDanger(unit) or dfhack.units.isGreatDanger(unit) then
+            table.insert(units, unit)
+        end
+    end
+
+    return units
+end
+
+local function getAllUnits()
+    local units = {}
+
+    for _, unit in pairs(df.global.world.units.active) do
+        table.insert(units, unit)
+    end
+
+    return units
+end
+
+UnitSyndromes = defclass(UnitSyndromes, widgets.Window)
+UnitSyndromes.ATTRS {
+    frame_title='Unit Syndromes',
+    frame={w=50, h=30},
+    resizable=true,
+    resize_min={w=43, h=20},
+}
+
+function UnitSyndromes:init()
+    local is_not_main_page = function()
+        return self.subviews.pages:getSelected() > 1
+    end
+
+    local previous_page = function()
+        self.subviews.pages:setSelected(self.subviews.pages:getSelected() - 1)
+    end
+
+    self:addviews{
+        widgets.Pages{
+            view_id = 'pages',
+            frame = {t = 0, l = 0, b = 2},
+            subviews = {
+                widgets.List{
+                    view_id = 'category',
+                    frame = {t = 0, l = 0},
+                    choices = {
+                        { text = "Dwarves", get_choices = getCitizens },
+                        { text = "Livestock", get_choices = getLivestock },
+                        { text = "Wild animals", get_choices = getWildAnimals },
+                        { text = "Hostile", get_choices = getHostiles },
+                        { text = "All units", get_choices = getAllUnits },
+                        { text = "All syndromes" },
+                    },
+                    on_submit = self:callback('showUnits'),
+                },
+                widgets.List{
+                    view_id = 'units',
+                    frame = {t = 0, l = 0},
+                    row_height = 3,
+                    on_submit = self:callback('showUnitSyndromes'),
+                },
+                widgets.List{
+                    view_id = 'unit_syndromes',
+                    frame = {t = 0, l = 0},
+                    on_submit = self:callback('showSyndromeEffects'),
+                },
+                widgets.WrappedLabel{
+                    view_id = 'syndrome_effects',
+                    frame = {t = 0, l = 0},
+                    text_pen = COLOR_WHITE,
+                    cursor_pen = COLOR_WHITE,
+                    text_to_wrap = '',
+                }
+            }
+        },
+        widgets.HotkeyLabel{
+            frame={l=0, b=0},
+            label='Back',
+            auto_width=true,
+            key='KEYBOARD_CURSOR_LEFT',
```

Comment:
This is displayed as a number since the keypad is also bound to this action. In other tools I render over the number with an arrow symbol. e.g. https://github.com/DFHack/scripts/blob/master/quickstart-guide.lua#L78-L82

Reply:
Note that you can also use `key='LEAVESCREEN'` to use Esc for "back", which other tools like `gui/gm-unit` do. This is perfectly acceptable. Since you deactivate the hotkey when you can no longer go back, the ZScreen code will still be able to dismiss the screen properly when the player hits Esc then.

Note if you use Esc/LEAVESCREEN, though, you should also have an onInput() function that interprets keys._R_MOUSE_DOWN to also call `previous_page` when `is_not_main_page()`

Reply:
That's a great call, had no idea that zscreen manages the esc logic. Good to know.